### ISSUE TITLE
SetArrayArgument action has incorrect type deduction

### ISSUE THREAD

**Issue by Warepire** [google/googletest#2874]
Using SetArrayArgument as an action on a mocked function returning non-void does not cause a compilation error if used directly in WillOnce / WillRepeatedly without DoAll and Return.

Other actions such as SetArgPointee does.

Seen both on 1.8.x and master.

I can probably provide a reproducer later if needed.

**Comment by Warepire** [comment]
```c++
#include "gmock/gmock.h"
#include "gtest/gtest.h"

#include <iostream>
#include <string>

using ::testing::_;
using ::testing::DoAll;
using ::testing::Return;
using ::testing::SetArgPointee;
using ::testing::SetArrayArgument;

class ISomeInterface {
 public:
  virtual ~ISomeInterface() = default;

  virtual unsigned int FuncThatTakesArrayArg(char* arr, size_t len) = 0;
  virtual unsigned int FuncThatTakesPointer(char* ptr) = 0;
};

class SomeInterfaceMock : public ISomeInterface {
 public:
  MOCK_METHOD2(FuncThatTakesArrayArg, unsigned int(char*, size_t));
  MOCK_METHOD1(FuncThatTakesPointer, unsigned int(char*));
};

class ClassUnderTest {
 public:
  ClassUnderTest(ISomeInterface *ifc) : ifc_(ifc) {
  }

  void MethodCalledByTest() {
    char array[10] = {};
    unsigned int ret = ifc_->FuncThatTakesArrayArg(array, sizeof(array));
    if (ret == 0 || ret >= sizeof(array))
      return;

    std::cout << "Array contains: " << array << std::endl;
  }

  void AnotherMethodCalledByTest() {
    char x = 'x';
    unsigned int ret = ifc_->FuncThatTakesPointer(&x);
    if (ret != 0)
      return;

    std::cout << "x contains " << x << std::endl;
  }

 private:
  ISomeInterface *ifc_;
};

class UnitTest : public ::testing::Test {
 public:
  void SetUp() override {
    cut_ = new ClassUnderTest(&mock_);
  }
  void TearDown() override {
    delete cut_;
  }
 protected:
  ClassUnderTest *cut_;
  SomeInterfaceMock mock_;
};

// Compiles fine, despite the lack of DoAll and Return
// On execution the stack becomes unbalanced and the test application crashes with segmentation fault:
TEST_F(UnitTest, SetArrayArgumentUsage) {
  std::string str = "humbug";
  EXPECT_CALL(mock_, FuncThatTakesArrayArg(_, 10))
    .WillRepeatedly(SetArrayArgument<0>(str.begin(), str.end()));

  cut_->MethodCalledByTest();
}

// Fixed version that doesn't break the stack:
//TEST_F(UnitTest, SetArrayArgumentCorrectUsage) {
//  std::string str = "humbug";
//  EXPECT_CALL(mock_, FuncThatTakesArrayArg(_, 10))
//    .WillRepeatedly(DoAll(SetArrayArgument<0>(str.begin(), str.end()), Return(5)));
//
//  cut_->MethodCalledByTest();
//}

// Doesn't compile fine, despite the same usage scenario as SetArrayArgument
TEST_F(UnitTest, SetArgumentPointeeUsage) {
  EXPECT_CALL(mock_, FuncThatTakesPointer(_))
    .WillRepeatedly(SetArgPointee<0>('g'));

  cut_->AnotherMethodCalledByTest();
}

// Fixed version that compiles:
//TEST_F(UnitTest, SetArgumentPointeeCorrectUsage) {
//  EXPECT_CALL(mock_, FuncThatTakesPointer(_))
//    .WillRepeatedly(DoAll(SetArgPointee<0>('g'), Return(0)));
//
//  cut_->AnotherMethodCalledByTest();
//}
```

Fails on current google test master, fails on latest googletest 1.8.x commit.
GCC 9.3.0
Compile options when building test: -std=c++17 -O2

**Comment by zhangxy988** [comment]
Thanks for reporting this issue. After some digging, I found that the fact that `SetArrayArgument` compiles is actually UB (Undefined Behavior). I was able to make this fail to compile in our internal toolchains and on MSVC on godbolt (https://gcc.godbolt.org/z/NZgYK4). The warning for clang/gcc to detect this should be `-Wreturn-type` but somehow it is not triggering on godbolt (I suspect they marked googletest as system headers and disabled warnings for system headers).
In summary, you should always specify `Return` action for any method that returns non-void. Whether this is caught by compiler depends on whether the warning `-Wreturn-type` is enabled in your toolchain configuration.

**Comment by zhangxy988** [comment]
As an added proof, https://gcc.godbolt.org/z/45ao4Q shows that if we copy the same action to user code, the warning is triggered.
